File: apps/api/src/larder/routers/meals.py
from typing import Literal
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Request, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from larder.auth.deps import CurrentUser, require_household
from larder.db.session import get_session
from larder.schemas.meals import FeedbackCreate, FeedbackOut, MealCreate, MealOut, MealPatch, MealsOut
from larder.services import meals as svc

router = APIRouter(prefix="/meals", tags=["meals"])


async def _out(session: AsyncSession, household_id: UUID, meal) -> MealOut:
    summary = await svc.feedback_summary(session, household_id, [meal.id])
    return MealOut.from_model(meal, summary.get(meal.id))


@router.get("", response_model=MealsOut)
async def list_meals(
    query: str | None = Query(default=None, max_length=80),
    meal_type: str | None = Query(default=None, max_length=32),
    source: Literal["user", "generated", "all"] = "user",
    user: CurrentUser = Depends(require_household),
    session: AsyncSession = Depends(get_session),
) -> MealsOut:
    meals = await svc.list_meals(session, user.household_id, query, meal_type, source)
    summaries = await svc.feedback_summary(session, user.household_id, [m.id for m in meals])
    return MealsOut(meals=[MealOut.from_model(m, summaries.get(m.id)) for m in meals])


@router.post("", response_model=MealOut, status_code=status.HTTP_201_CREATED)
async def create_meal(
    body: MealCreate,
    request: Request,
    user: CurrentUser = Depends(require_household),
    session: AsyncSession = Depends(get_session),
) -> MealOut:
    meal = await svc.create_meal(session, request.app.state.llm, user.household, user.profile.id, body)
    return await _out(session, user.household_id, meal)


@router.get("/{meal_id}", response_model=MealOut)
async def get_meal(
    meal_id: UUID, user: CurrentUser = Depends(require_household), session: AsyncSession = Depends(get_session)
) -> MealOut:
    return await _out(session, user.household_id, await svc.get_meal(session, user.household_id, meal_id))


@router.patch("/{meal_id}", response_model=MealOut)
async def patch_meal(
    meal_id: UUID,
    patch: MealPatch,
    user: CurrentUser = Depends(require_household),
    session: AsyncSession = Depends(get_session),
) -> MealOut:
    meal = await svc.get_meal(session, user.household_id, meal_id)
    return await _out(session, user.household_id, await svc.update_meal(session, user.household_id, meal, patch))


@router.delete("/{meal_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_meal(
    meal_id: UUID, user: CurrentUser = Depends(require_household), session: AsyncSession = Depends(get_session)
) -> Response:
    meal = await svc.get_meal(session, user.household_id, meal_id)
    await svc.delete_meal(session, meal)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{meal_id}/enrich", response_model=MealOut)
async def enrich(
    meal_id: UUID,
    request: Request,
    user: CurrentUser = Depends(require_household),
    session: AsyncSession = Depends(get_session),
) -> MealOut:
    meal = await svc.get_meal(session, user.household_id, meal_id)
    meal = await svc.re_enrich(session, request.app.state.llm, user.household, meal)
    return await _out(session, user.household_id, meal)


@router.post("/{meal_id}/feedback", response_model=FeedbackOut, status_code=status.HTTP_201_CREATED)
async def feedback(
    meal_id: UUID,
    body: FeedbackCreate,
    user: CurrentUser = Depends(require_household),
    session: AsyncSession = Depends(get_session),
) -> FeedbackOut:
    meal = await svc.get_meal(session, user.household_id, meal_id)
    summary = await svc.add_feedback(
        session, user.household_id, user.profile.id, meal, body.kind, body.plan_entry_id, body.comment
    )
    return FeedbackOut(feedback=summary)

File: packages/api-client/src/hooks/meals.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { unwrap, type Schemas } from "../client";
import { keys } from "../keys";
import { useApi } from "../provider";

export type MealOut = Schemas["MealOut"];
export type MealCreate = Schemas["MealCreate"];
export type MealPatch = Schemas["MealPatch"];
export type FeedbackKind = Schemas["FeedbackCreate"]["kind"];

export interface MealListParams {
  query?: string;
  meal_type?: string;
  source?: "user" | "generated" | "all";
}

export function useMeals(params: MealListParams = {}, enabled = true) {
  const api = useApi();
  return useQuery({
    queryKey: keys.meals(params as Record<string, string | undefined>),
    queryFn: async () => unwrap(await api.raw.GET("/api/v1/meals", { params: { query: params } })),
    enabled,
  });
}

export function useMeal(id: string | undefined) {
  const api = useApi();
  return useQuery({
    queryKey: keys.meal(id ?? ""),
    queryFn: async () => unwrap(await api.raw.GET("/api/v1/meals/{meal_id}", { params: { path: { meal_id: id! } } })),
    enabled: !!id,
  });
}

function useInvalidateMeals() {
  const qc = useQueryClient();
  return (id?: string) => {
    void qc.invalidateQueries({ queryKey: ["meals"] });
    if (id) void qc.invalidateQueries({ queryKey: keys.meal(id) });
    void qc.invalidateQueries({ queryKey: keys.plans });
  };
}

export function useCreateMeal() {
  const api = useApi();
  const invalidate = useInvalidateMeals();
  return useMutation({
    mutationFn: async (body: MealCreate) => unwrap(await api.raw.POST("/api/v1/meals", { body })),
    onSuccess: () => invalidate(),
  });
}

export function useUpdateMeal() {
  const api = useApi();
  const invalidate = useInvalidateMeals();
  return useMutation({
    mutationFn: async (vars: { id: string; patch: MealPatch }) =>
      unwrap(await api.raw.PATCH("/api/v1/meals/{meal_id}", { params: { path: { meal_id: vars.id } }, body: vars.patch })),
    onSuccess: (_d, vars) => invalidate(vars.id),
  });
}

export function useDeleteMeal() {
  const api = useApi();
  const invalidate = useInvalidateMeals();
  return useMutation({
    mutationFn: async (id: string) => unwrap(await api.raw.DELETE("/api/v1/meals/{meal_id}", { params: { path: { meal_id: id } } })),
    onSuccess: () => invalidate(),
  });
}

export function useEnrichMeal() {
  const api = useApi();
  const invalidate = useInvalidateMeals();
  return useMutation({
    mutationFn: async (id: string) =>
      unwrap(await api.raw.POST("/api/v1/meals/{meal_id}/enrich", { params: { path: { meal_id: id } } })),
    onSuccess: (_d, id) => invalidate(id),
  });
}

export function useFeedback() {
  const api = useApi();
  const invalidate = useInvalidateMeals();
  return useMutation({
    mutationFn: async (vars: { mealId: string; kind: FeedbackKind; plan_entry_id?: string; comment?: string }) =>
      unwrap(
        await api.raw.POST("/api/v1/meals/{meal_id}/feedback", {
          params: { path: { meal_id: vars.mealId } },
          body: { kind: vars.kind, plan_entry_id: vars.plan_entry_id, comment: vars.comment },
        }),
      ),
    onSuccess: (_d, vars) => invalidate(vars.mealId),
  });
}
